protocols: Implement Direct Connection Upgrade through Relay (dcutr)

mxinden opened the pull request:
This pull request implements the [Direct Connection Upgrade through Relay](https://github.com/libp2p/specs/pull/173) specification. For the greater picture please see https://github.com/libp2p/rust-libp2p/issues/2052 and https://github.com/libp2p/specs/issues/312.

**Status quo**

It builds on top of https://github.com/libp2p/rust-libp2p/pull/2059 and https://github.com/libp2p/rust-libp2p/pull/2066. Please post review comments related to circuit relay v2 and simultaneous open on https://github.com/libp2p/rust-libp2p/pull/2059 and https://github.com/libp2p/rust-libp2p/pull/2066.

mxinden committed:
protocols/relay: Implement circuit relay v2 protocol

This commit adds an implementation for the circuit relay v2 protocol to
be used as a relay server, i.e. it supports incoming HOP requests and
outgoing STOP requests. Future commits will add support for clients,
i.e. outgoing HOP requests and incoming STOP requests.

The existing circuit relay v1 protocol implementation is moved to
protocols/relay/src/v1.

mxinden committed:
misc/multistream-select: Ignore simultaneous open 'iamclient'

mxinden committed:
protocols/relay: Ensure connections of HOP connect are kept alive

mxinden committed:
protocols/relay: Improve documentation

mxinden committed:
misc/multistream-select: Implement simultaneous open extension

From the multistream-select 1.0 simultaneous open protocol extension
specification:

> In order to support direct connections through NATs with hole
punching, we need to account for simultaneous open. In such cases, there
is no single initiator and responder, but instead both peers act as
initiators. This breaks protocol negotiation in multistream-select,
which assumes a single initator.

> This draft proposes a simple extension to the multistream protocol
negotiation in order to select a single initator when both peers are
acting as such.

See https://github.com/libp2p/specs/pull/196/ for details.

This commit implements the above specification, available via
`Version::V1SimOpen`.

mxinden committed:
protocols/relay: Implement v2 client logic

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
protocols/relay: Handle dial failure

mxinden committed:
protocols/relay: Reuse connection

mxinden committed:
protocols/relay: Rename Connection to RelayedConnection

mxinden committed:
protocols/relay: Update transport doc examples

mxinden committed:
protocols/relay: Pass relay addr to transport

mxinden committed:
protocols/relay: Implement inbound stop denial

mxinden committed:
protocols/relay: Renew reservations

mxinden committed:
protocols/relay: Handle invalid expiration in the past

mxinden committed:
protocols/relay: Handle in and outbound failure

mxinden committed:
protocols/relay: Implement client handler keep alive

mxinden committed:
protocols/relay: Handle handler listener closed channel

mxinden committed:
protocols/relay: Handle handler to listener failure

mxinden committed:
protocols/relay: Return all new listener addresses

mxinden committed:
protocols/dcutr: Implement Direct Connection Upgrade through Relay

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
Merge branch 'libp2p/master' into dcutr

mxinden committed:
protocols/dcutr: Carry observed addresses in Connect

mxinden committed:
protocols/dcutr: Reply to Syn

mxinden committed:
protocols/dcutr: Dial as initiator

mxinden committed:
core/: Integrate Simultaneous Open extension

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
protocols/relay/v2: Update to latest protobuf definition

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
protocols/dcutr/examples: Add client

mxinden committed:
Revert "misc/multistream-select: Ignore simultaneous open 'iamclient'"

This reverts commit 125e3c3e5740fb9334a3ccc3e9598f2479fbd635.

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
Merge branch 'sim-open' into dcutr

mxinden committed:
core/src/transport/upgrade: Make DialFuture aware of SimOpenRole

mxinden committed:
core/src/transport: Use Transport::and_then for Authenticated::apply

mxinden committed:
core/: Clean type structure

mxinden committed:
core/: Enforce upgrade version at compile time

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
misc/multistream-select: Document V1SimOpen

mxinden committed:
*: Rename V1SimOpen to V1SimultaneousOpen

mxinden committed:
misc/multistream-select: Document SimOpenRole

mxinden committed:
*: Rename SimOpenRole to Role

mxinden committed:
misc/multistream-select: Document reponder role process

mxinden committed:
misc/multistream-select: Bump version and add changelog entry

mxinden committed:
core/CHANGELOG: Add entry

mxinden committed:
core/src/upgrade: Assert Initiator when not using SimOpen

mxinden committed:
core/upgrade/apply: Document different versions

mxinden committed:
misc/multistream-select: Derive Eq for Role

mxinden committed:
*: Fix documentation links

mxinden committed:
misc/multistream-select: Fix doc link

mxinden committed:
src/lib: Call upgrade without Version

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
protocols/relay/v2: Report back to transport

mxinden committed:
protocols/relay/v2: Disconnect when stop protocol not supported

mxinden committed:
protocols/relay/v2: Document max_duration not exceed u32::MAX

mxinden committed:
protocols/relay/v2: Don't append p2p-circuit as relay

mxinden committed:
protocols/relay/v2: Implement rate limiter

mxinden committed:
protocols/relay/v2: Document caveats on rate limiter with high volume

mxinden committed:
protocols/relay: Prevent possible false positive in quickcheck

mxinden committed:
protocols/relay: Reword Prost error message

mxinden committed:
protocols/relay: Allow users to specify generic rate limiters

mxinden committed:
protocols/relay: Move rate limiting logic into module

mxinden committed:
protocols/relay: Prevent reservation and connection over relayed conn

mxinden committed:
protocols/relay: Add circuit src rate limiting

mxinden committed:
protocols/relay/v2: Simplify example

mxinden committed:
protocols/relay: Add myself to authors

mxinden committed:
protocols/relay: Use thiserror

mxinden committed:
protocols/relay/v2: Set rate limits

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
protocols/relay: Use wasm_timer::Instant

mxinden committed:
protocols/relay/v2: Apply clippy suggestions

mxinden committed:
protocols/relay: Fix intra doc link

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
Merge branch 'sim-open' into dcutr

mxinden committed:
protocols/dcutr/tests: Fix connect test

mxinden committed:
protocols/dcutr: Run cargo fix

mxinden committed:
protocols/dcutr: Attempt direct connection upgrade as listener only

mxinden committed:
protocols/dcutr: Emit event when attempting direct connection upgrade

mxinden committed:
protocols/dcutr: Emit event when remote initiates direct connection up

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
protocols/relay: Fix clippy warnings

mxinden committed:
protocols/dcutr: Extend with TODOs

mxinden committed:
swarm/src/behaviour: Update DialAddress doc comment

With 45f07bf8639fd9056e4ffe52298ca113a0308951 `Network::dial` accepts a
`Multiaddr` with a `PeerId`. With that in mind the doc comment on
`NetworkBehaviourAction::DialAddress` is outdated.

mxinden committed:
swarm/src/protocols_handler: Add EitherHandler

Add implementation of a ProtocolsHandler that represents either of two
ProtocolsHandler implementations.

mxinden committed:
protocols/dcutr: Deny upgrades on non-relayed connections

mxinden committed:
protocols/dcutr: Properly expect size of data container

mxinden committed:
protocols/dcutr: Enforce maximum message size on incoming msgs

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
Merge commit '008561283ed0a0530d449fdc7ff4dc9f3ab9ce73' into sim-open

mxinden committed:
*: Format with rustfmt

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
Merge branch 'sim-open' into dcutr

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
Merge branch 'sim-open' into dcutr

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
protocols/dcutr: Tell handler what it is going to be used for

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
protocols/dcutr: Retry on upgrade failure

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
misc/metrics: Add basic instrumentation for libp2p-relay

r-zig commented on misc/multistream-select/src/protocol.rs:
```diff
@@ -249,7 +303,7 @@ impl<R> MessageIO<R> {
     }
 
-    /// Drops the [`MessageIO`] resource, yielding the underlying I/O stream.
+    /// Draops the [`MessageIO`] resource, yielding the underlying I/O stream.
```
spelling -> Drops?

r-zig commented on protocols/dcutr/examples/client.rs:
```diff
@@ -0,0 +1,178 @@
+        panic!();
+    }
+
+    if let (Some(mode), Some(addr)) = (std::env::args().nth(4), std::env::args().nth(5)) {
```
Maybe we can change to structopt with description for each parameter as we did in the relay_v1
and then we can write some example of 3 nodes settings together?

mxinden replied:
Yes for sure, that would be nice. This is a hack at most at the moment. Want to file a pull request against my branch on my fork?

r-zig replied:
Yes, I will do.

r-zig committed:
protocols/relay/examples: Structure command line args

- Allow deterministic peer identity.
- Choose between ipv6 and ipv4.

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
Merge branch 'libp2p/master' into sim-open

mxinden committed:
Merge branch 'sim-open' into dcutr

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
protocols/dcutr: Report successful connection upgrade

mxinden committed:
protocols/dcutr/examples: Use structopts

mxinden committed:
protocols/dcutr: Do not panic on connection closing

mxinden committed:
protocols/relay: Return NetworkBehaviourAction::NotifyHandler right away

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
protocols/dcutr/src/behaviour: Do not send relayed addresses

mxinden committed:
protocols/dcutr/examples: Log on info level

mxinden committed:
protocols/relay: Run rust fmt

mxinden committed:
protocols/relay/src/v2/relay: Accept mutable config

mxinden committed:
protocols/relay/src/v2/relay: Accept mutable config

mxinden committed:
misc/multistream-select/src/protocol.rs: Fix typo

mxinden committed:
Merge branch 'mxinden/relay-v2' into dcutr

mxinden committed:
protocols/dcutr/build: Run rust fmt

mxinden committed:
protocols/dcutr/src/behaviour: Remove outdated comment

mxinden committed:
misc/multistream-select: Supress needless collect warning

mxinden committed:
Merge branch 'mxinden/sim-open' into dcutr

mxinden committed:
protocols/dcutr/: Do not attempt upgrade if direct connection exists

mxinden committed:
protocols/dcutr/src/handler: Remove DirectConnection Prototype variant

mxinden committed:
protocols/dcutr: Deny incoming substreams as dialer

mxinden committed:
protocols/dcutr/src/handler: Split into relayed and direct

mxinden committed:
protocols/dcutr: Don't abbreviate negotiated with neg

mxinden committed:
protocols/dcutr/src/protocol: Use correct protocol name

mxinden committed:
protocols/dcutr/src/message.proto: Make type field required

mxinden committed:
protocols/dcutr/src/protocol: Use thiserror

mxinden committed:
protocols/dcutr/src/handler/relayed: Refine keep alive handling

mxinden committed:
protocols/dcutr/src/protocol: Split into outbound and inbound

mxinden committed:
protocols/dcutr/src/handler/relayed: Handle stream upgrade errors

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
Merge branch 'master' into relay-v2

mxinden committed:
Merge branch 'master' into dcutr

mxinden committed:
Merge branch 'relay-v2' into dcutr

canewsin commented on protocols/relay/src/v2.rs:
```diff
@@ -0,0 +1,41 @@
+
+pub mod client;
+mod copy_future;
+mod protocol;
```
Make protocol public mod or Export 
UpgradeError in inbound_hop/inbound_hop and two other modules

canewsin replied:
@mxinden

MarcoPolo committed:
Refactor suggestion

mxinden committed:
protocols/relay/src/v2/client: Return NetworkBehaviourAction on ListenReq

mxinden committed:
protocols/relay/src/v2/client/handler: Check status of lend out substreams

mxinden committed:
protocols/relay/src/v2/client: Use void::Void for drop_notifer

mxinden committed:
protocols/relay/src/v2/client/handler: Log dropped oneshot Sender to transport

mxinden committed:
protocols/relay/src/v2/client/handler: Await send call to transport listener

mxinden committed:
protocols/relay/src/v2/client/handler: Remove unnecessary boxing

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

mxinden committed:
protocols/relay/v2: Use AcceptReservationReqBuilder concept

thomaseizinger committed:
protocols/relay/src/v2/client: Introduce Reservation::None

This allows us to put more logic on the actual `Reservation` type,
making the implementation of the handler slightly simpler.

mxinden committed:
protocols/relay/src/v2/client: Extend unreacheable! text

mxinden committed:
protocols/relay/src/v2/: Document pending_error and send_error_futs

mxinden committed:
protocols/relay/src/v2: Use OrTransport

mxinden committed:
protocols/relay/src/v2: Fix typo

mxinden committed:
protocols/relay/src/v2: Renew reservation after 3/4 of expiration

mxinden committed:
protocols/relay: Capitalize I in NoAddressesInReservation

mxinden committed:
misc/metrics: Derive Debug for event types

mxinden committed:
protocols/relay/src/v2/client/handler: Consistent mpsc and oneshot import

mxinden committed:
protocols/relay/src/v2/client/transport: Loop on from_behaviour

Previous version was (a) confusing due to assuming `queued_new_addresses` to be
empty and (b) error prone in the case where `addrs` is emtpy and thus no `Waker`
being registered.

mxinden committed:
protocols/relay: Fix intra doc link

mxinden committed:
protocols/relay/: Allow large enum variant

While a good lint in itself, I don't think we need to optimize these enums
without previously benchmarking its impact.

mxinden committed:
protocols/relay/src/v2/client/transport: Remove destination address

Don't include destination address when establishing relayed connection in doc
example.

mxinden committed:
protocols/relay/src/v2: Don't override NetworkBehaviour methods with default

mxinden committed:
protocols/relay/src/v2/client: Pass event at handler initialization

Instead of creating a new connection and only once established pass a command to
said connection, provide the command early on to the handler, thus not requiring
to buffer the command within the `NetworkBehaviour` implementation.

mxinden committed:
protocols/relay/src/v2: Use lent instead of lend

mxinden committed:
protocols/relay/v2: Use Renewing instead of Renewal

mxinden committed:
protocols/relay/src/v2: Fix intra doc link

mxinden committed:
Merge branch 'libp2p/master' into relay-v2

MarcoPolo committed:
protocols/relay/src/v2/client/handler: Refactor Reservation::poll

mxinden committed:
protocols/relay/src/v2/client/handler: Inject event at creation time

mxinden committed:
protocols/relay/src/v2/client/handler.rs: Fix typo

Co-authored-by: Elena Frank <57632201+elenaf9@users.noreply.github.com>

mxinden committed:
protocols/relay/src/v2/client/transport.rs: Make doc examples consistent

Co-authored-by: Elena Frank <57632201+elenaf9@users.noreply.github.com>

mxinden committed:
protocols/relay/src/v2: Stress required usage of other Transport

mxinden committed:
protocols/relay/src/v2/protocol/inbound_stop.rs: Remove src_peer_id pub(crate)

Co-authored-by: Elena Frank <57632201+elenaf9@users.noreply.github.com>

mxinden committed:
protocols/relay/src/v2/relay/handler: Remove wrong unused variable assignment

Co-authored-by: Elena Frank <57632201+elenaf9@users.noreply.github.com>

mxinden committed:
protocols/relay/src/v2/client: Garbage collect unconnected nodes

mxinden committed:
protocols/relay/src/v2: Prevent nested relay on incoming relayed conn

mxinden committed:
protocols/relay/tests/v2: Test new reservation replacing old

mxinden committed:
protocols/relay/src/v2: Close listener when channel from behaviour dropped

mxinden committed:
protocols/relay/src/v2: Expose connection limit via OutEvent

mxinden committed:
protocols/relay/v2: Use Option::transpose

mxinden committed:
protocols/relay/src/v2/client: Queue Event and not Action

mxinden committed:
protocols/relay/src/v2/client: Remove loop in NetworkBehaviour::poll

mxinden committed:
protocols/relay/src/v2/client/: Refactor parse_relayed_multiaddr

mxinden committed:
protocols/relay/src/v2: Use ready! in RelayListener as Stream

mxinden committed:
protocols/relay/src/v2: Remove unnecessary else

mxinden committed:
protocols/relay/src/v2/protocol: Use Vec::with_capacity

mxinden committed:
protocols/relay/src/v2: Separate error paths and happy path

mxinden committed:
protocols/relay/src/v2/copy_future.rs: Add unit tests

- Copying from a to b and vice versa.
- Checking max bytes.
- Checking max duration.

mxinden committed:
protocols/relay/src/v2/relay: Remove unnecessary unwrap_or

mxinden committed:
Merge remote-tracking branch 'libp2p/master' into relay-v2

mxinden committed:
protocols/relay/src/v2: Use ConnectedPoint::is_relayed

mxinden committed:
Merge branch 'relay-v2' into dcutr

mxinden committed:
protocols/dcutr: Use ConnectedPoing::is_relayed

mxinden committed:
protocols/dcutr: Use ConnectedPoing::is_relayed 2

mxinden committed:
protocols/dcutr/: Remove unnecessary pub export of handler

mxinden committed:
protocols/dcutr: Use ActionBuilder construct

See 112395275d9a9838fe802586090e1e2777e15f91 for similar change in libp2p-relay.

mxinden committed:
protocols/dcutr: Refactor error handling

See https://github.com/mxinden/rust-libp2p/pull/7 for similar change in
libp2p-relay.

mxinden committed:
misc/metrics: Add support for libp2p-dcutr

mxinden committed:
protocols/dcutr/Cargo: Update rust version
